### PR TITLE
Job db

### PR DESCRIPTION
**Pull request by bertsky**
Builds on #29, adds using the database for job monitoring as well – depends on https://github.com/slub/ocrd_manager/pull/62

### CODE REVIEW

**Comment by bertsky** [comment]
Minimal diff (against #29 directly) is [here](https://github.com/SvenMarcus/ocrd_monitor/pull/1)

**Comment by bertsky** [comment]
To be able to [kill active jobs](https://github.com/slub/ocrd_monitor/pull/24/commits/d5cd6330a3d3be868b2d76bc26bb12b47b599203) or start new ones, we would still need to add a minimal web service in the Manager. (Ideally written in Python and re-usable for the core functionalities of both process_images.sh and process_mets.sh ...)

**Comment by bertsky** [comment]
> To be able to [kill active jobs](https://github.com/slub/ocrd_monitor/pull/24/commits/d5cd6330a3d3be868b2d76bc26bb12b47b599203)

Done in 89aed9e, dependent on recent changes in https://github.com/slub/ocrd_manager/pull/62

File: ocrdmonitor/server/templates/jobs.html.j2 [hunk 1]
```diff
@@ -11,8 +11,7 @@
 {% block content %}
 <script>
     function killjob(pid) {
-    /* todo: log into the Manager and kill this process ... */
-        window.alert("Not implemented yet")
+        console.log(fetch(`http://ocrd-manager:4004/cancel_job/${pid}`));
```
Comment (bertsky) [inline ocrdmonitor/server/templates/jobs.html.j2:14]: @SvenMarcus says this must be an `await` too

**Comment by SvenMarcus** [comment]
I have rebased the branch onto my work on the browser DB and implemented the necessary refactorings to bring all these changes together.

**Comment by SvenMarcus** [comment]
For some reason, this PR is missing a commit, despite being pushed to the job-db branch. Mypy fails, although the attribute name has already been fixed in the latest commit. [Link to code line](https://github.com/slub/ocrd_monitor/blob/82727582cdba7d45701e7d239aded34b0095567d/ocrdmonitor/environment.py#L33)

**Review by SvenMarcus — APPROVED** [review]

**Comment by SvenMarcus** [comment]
Adding another commit seems to have resolved the issue.

**Comment by markusweigelt** [comment]
@SvenMarcus When I start the monitor via the base repo I get the following error. Is there something wrong with the configuration or is this still a problem in the monitor.

```
Traceback (most recent call last):
  File "/usr/local/bin/uvicorn", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/uvicorn/main.py", line 416, in main
    run(
  File "/usr/local/lib/python3.11/site-packages/uvicorn/main.py", line 587, in run
    server.run()
  File "/usr/local/lib/python3.11/site-packages/uvicorn/server.py", line 61, in run
    return asyncio.run(self.serve(sockets=sockets))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/uvicorn/server.py", line 68, in serve
    config.load()
  File "/usr/local/lib/python3.11/site-packages/uvicorn/config.py", line 467, in load
    self.loaded_app = import_from_string(self.app)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/uvicorn/importer.py", line 21, in import_from_string
    module = importlib.import_module(module_str)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/usr/local/ocrd-monitor/ocrdmonitor/main.py", line 5, in <module>
    settings = Settings()
               ^^^^^^^^^^
  File "pydantic/env_settings.py", line 40, in pydantic.env_settings.BaseSettings.__init__
  File "pydantic/main.py", line 341, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for Settings
monitor_db_connection_string
  field required (type=value_error.missing)
```

ATM I checked out the open pr branches in the submodules without testing the monitor individually.

**Comment by SvenMarcus** [comment]
> @SvenMarcus When I start the monitor via the base repo I get the following error. Is there something wrong with the configuration or is this still a problem in the monitor.
> 
> ```
> Traceback (most recent call last):
>   File "/usr/local/bin/uvicorn", line 8, in <module>
>     sys.exit(main())
>              ^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
>     return self.main(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1078, in main
>     rv = self.invoke(ctx)
>          ^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/click/core.py", line 783, in invoke
>     return __callback(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/uvicorn/main.py", line 416, in main
>     run(
>   File "/usr/local/lib/python3.11/site-packages/uvicorn/main.py", line 587, in run
>     server.run()
>   File "/usr/local/lib/python3.11/site-packages/uvicorn/server.py", line 61, in run
>     return asyncio.run(self.serve(sockets=sockets))
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/asyncio/runners.py", line 190, in run
>     return runner.run(main)
>            ^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/asyncio/runners.py", line 118, in run
>     return self._loop.run_until_complete(task)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
>     return future.result()
>            ^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/uvicorn/server.py", line 68, in serve
>     config.load()
>   File "/usr/local/lib/python3.11/site-packages/uvicorn/config.py", line 467, in load
>     self.loaded_app = import_from_string(self.app)
>                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/uvicorn/importer.py", line 21, in import_from_string
>     module = importlib.import_module(module_str)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
>   File "<frozen importlib._bootstrap_external>", line 940, in exec_module
>   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
>   File "/usr/local/ocrd-monitor/ocrdmonitor/main.py", line 5, in <module>
>     settings = Settings()
>                ^^^^^^^^^^
>   File "pydantic/env_settings.py", line 40, in pydantic.env_settings.BaseSettings.__init__
>   File "pydantic/main.py", line 341, in pydantic.main.BaseModel.__init__
> pydantic.error_wrappers.ValidationError: 1 validation error for Settings
> monitor_db_connection_string
>   field required (type=value_error.missing)
> ```
> 
> ATM I checked out the open pr branches in the submodules without testing the monitor individually.

The meta repository doesn't contain the necessary configuration yet. The monitor has some renamed and new variables (see docker-compose.yml).

**Comment by markusweigelt** [comment]
Ok i will check this. I thought that all changes regarding Monitor adjustments were made in this [Base Repository PR](https://www.google.com/search?q=https%3A%2F%2Fgithub.com%2Fslub%2Focrd_kitodo%2Fpull%2F89](https://github.com/slub/ocrd_kitodo/pull/8)

**Comment by markusweigelt** [comment]
In `docker-compose.yml`  we extend from the submodule Docker Compose file. The `add-database` branch of @bertsky is linked to the changes of this PR 

https://github.com/slub/ocrd_kitodo/blob/745f5d7b686eaf067e85a4d7c55a57e515f30e6d/docker-compose.yml#L13C1-L19
and 
https://github.com/slub/ocrd_kitodo/blob/745f5d7b686eaf067e85a4d7c55a57e515f30e6d/docker-compose.yml#L26-L37

So the `MONITOR_DB_CONNECTION` env should be available.

**Comment by markusweigelt** [comment]
Docker inspect of not running Monitor container outputs following env`s 

```
"Env": [
    "MONITOR_PORT_LOG=8088",
    "CONTROLLER=ocrd-controller:22",
    "MONITOR_DB_CONNECTION=mongodb://root:root_password@ocrd-database:27017",
    "PATH=/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
    "LANG=C.UTF-8",
    "GPG_KEY=A035C8C19219BA821ECEA86B64E628F8D684696D",
    "PYTHON_VERSION=3.11.4",
    "PYTHON_PIP_VERSION=23.1.2",
    "PYTHON_SETUPTOOLS_VERSION=65.5.1",
    "PYTHON_GET_PIP_URL=https://github.com/pypa/get-pip/raw/0d8570dc44796f4369b652222cf176b3db6ac70e/public/get-pip.py",
    "PYTHON_GET_PIP_SHA256=96461deced5c2a487ddc65207ec5a9cffeca0d34e7af7ea1afc470ff0d746207"
],
```

**Comment by markusweigelt** [comment]
Problem solved with renaming `DB_CONNECTION_STRING` to `MONITOR_DB_CONNECTION_STRING` and `DB_CONNECTION` to `MONITOR_DB_CONNECTION` in `init.sh`.

**Comment by SvenMarcus** [comment]
Took me ages, but the settings finally work with pydantic v2. Maybe I'm using it wrong, but `pydantic-settings` for v2 seems way more unintuitive than the previous version, so we should probably keep it in mind in case we can simplify some things.

**Review by markusweigelt — APPROVED** [review]

**Comment by markusweigelt** [comment]
A small thing caught my attention for which I have created an issue. #31

**Pull request merged**